### DISCUSSION TITLE
"less" or "more" words

### DISCUSSION THREAD

**Discussion by mcctuxic** [tabemann/zeptoforth#121] [Ideas]
Hi,

the command "words" list all currently available words - starting with those defined last. Since Zeptoforth is a "not so small Forth" the list of words is long. Most of the time the top will be pushed out of the boundaries of the terminal - but the top of the list is the part, which I think most developers are interested in the most. Solultion: 
"Use the scroll wheel, Luke!"

Would it be possible without too much hassle to implement "less" or "more" functionality in the word "words"...
I mean: The UNIX tools "more" (which is most often "less" these days at least with Linux) fill one page of output and
wait for a key to be pressed. Is it a "q" it will quit and any other key will result in another page of printed output.

This way one can grasp the top of the list of words instantly.

only my two cent...your currency may vary ;)

**Comment by tabemann** [comment]
On Fri, Jul 26, 2024, 13:17 mcctuxic ***@***.***> wrote:

> Hi,
>
> the command "words" list all currently available words - starting with
> those defined last. Since Zeptoforth is a "not so small Forth" the list of
> words is long. Most of the time the top will be pushed out of the
> boundaries of the terminal - but the top of the list is the part, which I
> think most developers are interested in the most. Solultion:
> "Use the scroll wheel, Luke!"
>
> Would it be possible without too much hassle to implement "less" or "more"
> functionality in the word "words"...
> I mean: The UNIX tools "more" (which is most often "less" these days at
> least with Linux) fill one page of output and
> wait for a key to be pressed. Is it a "q" it will quit and any other key
> will result in another page of printed output.
>
> This way one can grasp the top of the list of words instantly.
>
> only my two cent...your currency may vary ;)
>

A 'more' approach could definitely be done. A 'less' approach would be
harder as the flash and RAM dictionaries are structured as singly linked
lists. Thanks for the suggestion!

(Sorry for my belated response - my webmail categorized your email
differently than it normally does so I missed it.)

Travis

>

**Comment by mcctuxic** [comment]
On 07/29 01:40, tabemann wrote:
> On Fri, Jul 26, 2024, 13:17 mcctuxic ***@***.***> wrote:
> 
> > Hi,
> >
> > the command "words" list all currently available words - starting with
> > those defined last. Since Zeptoforth is a "not so small Forth" the list of
> > words is long. Most of the time the top will be pushed out of the
> > boundaries of the terminal - but the top of the list is the part, which I
> > think most developers are interested in the most. Solultion:
> > "Use the scroll wheel, Luke!"
> >
> > Would it be possible without too much hassle to implement "less" or "more"
> > functionality in the word "words"...
> > I mean: The UNIX tools "more" (which is most often "less" these days at
> > least with Linux) fill one page of output and
> > wait for a key to be pressed. Is it a "q" it will quit and any other key
> > will result in another page of printed output.
> >
> > This way one can grasp the top of the list of words instantly.
> >
> > only my two cent...your currency may vary ;)
> >
> 
> A 'more' approach could definitely be done. A 'less' approach would be
> harder as the flash and RAM dictionaries are structured as singly linked
> lists. Thanks for the suggestion!
> 
> (Sorry for my belated response - my webmail categorized your email
> differently than it normally does so I missed it.)
> 
> Travis
> 
> >

Sorry...I wrote confusing stuff while trying to be funny in a foreign
language... :)

I just played around with the names of the Linux tools
"more" and "less". There is no need for a scroll back function IMHO.
I would be nice to have a pager function, which stops after each page of
words, can be quitted (word? looks strange somehow) then or instructed
to scroll all words in one go then (current behaviour) or inspect the
next page.

No need to be sorry, Travis!!! You you are creating updates, fixes and
releases and answers posts/emails in ___such a quick way___ !

Cheers!
Tuxic

**Comment by tabemann** [comment]
On Mon, Jul 29, 2024 at 10:09 PM mcctuxic ***@***.***> wrote:

> Sorry...I wrote confusing stuff while trying to be funny in a foreign
> language... :)
>
> I just played around with the names of the Linux tools
> "more" and "less". There is no need for a scroll back function IMHO.
> I would be nice to have a pager function, which stops after each page of
> words, can be quitted (word? looks strange somehow) then or instructed
> to scroll all words in one go then (current behaviour) or inspect the
> next page.
>
> No need to be sorry, Travis!!! You you are creating updates, fixes and
> releases and answers posts/emails in ___such a quick way___ !
>
> Cheers!
> Tuxic


I have made a new release, 1.7.1.1 (
https://github.com/tabemann/zeptoforth/releases/tag/v1.7.1.1), which
includes an important fix to IMPORT-FROM and BEGIN-IMPORTS-FROM and no
other changes... but I have subsequently implemented "more" style words
MORE-WORDS , MORE-WORDS-IN , MORE-LOOKUP , and MORE-LOOKUP-IN which output
words one screenful at a time, and which can be quit by pressing 'q' or
continued by pressing any other key. I did not implement a feature to go
all the way to the end because 1) that can be accomplished simply by
holding down a key and 2) it would have added a good bit of complexity to
my code. As mentioned this is not in a build yet, but you can get it
without making a whole new build by simply compiling more_words.fs (
https://github.com/tabemann/zeptoforth/blob/master/src/common/forth/more_words.fs
).

Travis

**Reply by mcctuxic** [reply]
OH _YEAH!_  T-H-A-N-K Y-O-U!

"My name is Gordon! ... _**FLASH**_ Gordon!" :)

My Pico is already updated....

**Reply by manes6969** [reply]
Thankks Travis !

**Comment by mcctuxic** [comment]
I added more words to my Zeptoforth by uploading more-words.fs to my Pico.. :)

more-words works like a charme! Great and realy helpful. I will the other new words later.

Before Zeptoforth I was experimenting with another forth on another well known 
chip. It was difficult to get it even to compile. Zeptoforth is really fantastic!
It is even possible to get a "sneak preview" version by including parts of an
upcoming version into the current one.

Man, how I wish more people would use Zeptoforth! It fits so well into the world 
of microcontroller programming!

"_You take the blue BASIC... the story ends, you wake up at your desk and program whatever you are told to. You take the red Zeptoforth... you stay in the land of the FORTH, and it will show you how deep the rabbit hole goes._"

...and I definetly DON'T play with all these more or less "philosophical" and gender related stuff, which was interpreted into 
this quote after The Matrix. I just like to alter the contents of quotes for another context. That's all!

**Comment by tabemann** [comment]
I have implemented updated versions of WORDS , LOOKUP , MORE-WORDS , and
MORE-LOOKUP that eliminate duplicated modules from the search order that is
displayed. However, because these changes affect basic.fs in addition to
more_words.fs I have only committed them to the devel branch and not to the
master branch as I do not have enough time tonight to put together another
build (and I do not want to break the version of more_words.fs in the
master branch, which currently does not require any changes to basic.fs).
Note that the legacy LOOKUP is currently broken because it leaves garbage
on the stack; I have a fix for this which will also be in the next release.

Travis

TravisMessage ID:
***@***.***>

>
